torchvision/ops/boxes.py

```diff
@@ -311,6 +311,50 @@ def generalized_box_iou(boxes1: Tensor, boxes2: Tensor) -> Tensor:
     return iou - (areai - union) / areai
 
 
+def distance_box_iou(boxes1: Tensor, boxes2: Tensor, eps: float = 1e-7) -> Tensor:
+    """
+    Return distance intersection-over-union (Jaccard index) between two sets of boxes.
+
+    Both sets of boxes are expected to be in ``(x1, y1, x2, y2)`` format with
+    ``0 <= x1 < x2`` and ``0 <= y1 < y2``.
+
+    Args:
+        boxes1 (Tensor[N, 4]): first set of boxes
+        boxes2 (Tensor[M, 4]): second set of boxes
+        eps (float, optional): small number to prevent division by zero. Default: 1e-7
+
+    Returns:
+        Tensor[N, M]: the NxM matrix containing the pairwise distance IoU values
+        for every element in boxes1 and boxes2
+    """
+    if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+        _log_api_usage_once(distance_box_iou)
+
+    boxes1 = _upcast(boxes1)
+    boxes2 = _upcast(boxes2)
+
+    inter, union = _box_inter_union(boxes1, boxes2)
+    iou = inter / union
+
+    lti = torch.min(boxes1[:, None, :2], boxes2[:, :2])
+    rbi = torch.max(boxes1[:, None, 2:], boxes2[:, 2:])
+
+    whi = _upcast(rbi - lti).clamp(min=0)  # [N,M,2]
+    diagonal_distance_squared = (whi[:, :, 0] ** 2) + (whi[:, :, 1] ** 2) + eps
+
+    # centers of boxes
+    x_p = boxes1[:, None, :2].sum() / 2
+    y_p = boxes1[:, None, 2:].sum() / 2
+    x_g = boxes2[:, :2].sum() / 2
+    y_g = boxes2[:, 2:].sum() / 2
```

Comment:
Hey @yassineAlouini , I think there is a problem with this implementation.  The calculation of centre of boxes is not correct acc to me. We should be adding up only `x1` `x2` and `y1` `y2`, [ref](https://github.com/facebookresearch/detectron2/blob/dfe8d368c8b7cc2be42c5c3faf9bdcc3c08257b1/detectron2/layers/losses.py#L49-L53) . 
But in  current implementation, we are adding `x1`,`y1` and `x2`, `y2`. (BBox shape is in form [x1,y1,x2,y2]) .

This can also be checked by calculating `distance_box_iou_loss` and `distance_box_iou` on a sample tensors. 


    import torch
    from torchvision.ops import distance_box_iou, distance_box_iou_loss

    box1 = torch.tensor([[-1, -1, 1, 1]], )
    box2 = torch.tensor([[0, 0, 1, 1]],)

    1-distance_box_iou(box1, box2)[0] == distance_box_iou_loss(box1, box2)


Last statement returns `False`. Ideally it should return `True`.

I suggest you to do following changes. 

```suggestion
    # centers of boxes
    x_p = (boxes1[:, 0] + boxes1[:, 2]) / 2
    y_p = (boxes1[:, 1] + boxes1[:, 3]) / 2
    x_g = (boxes2[:, 0] + boxes2[:, 2]) / 2
    y_g = (boxes2[:, 1] + boxes2[:, 3]) / 2
```

@datumbox,
Please correct me if I'm wrong.
Thanks.

Reply:
Good catch. I haven't yet reviewed the correctness of the implementation as we still discuss the structure/API. 

I think that's probably a typo and @yassineAlouini intended to write something like:
```
x_p = boxes1[:, 0::2].sum() / 2
y_p = boxes1[:, 1::2].sum() / 2
...
```

Reply:
Yes indeed, I think I went too quickly over this and thought that the bounding box was in the `x1x2y1y2` format. Thanks for pointing this out and your suggestions. :+1: